app/grandchallenge/profiles/adapters.py

```diff
@@ -44,6 +54,62 @@ def clean_email(self, email):
 
         return email
 
+    def pre_login(self, request, user, **kwargs):
+        # this is copied (and slightly adapted) from the a pending PR on django-allauth-2fa repo:
+        # https://github.com/valohai/django-allauth-2fa/pull/131
+
+        response = super().pre_login(request, user, **kwargs)
+        if response:
+            return response
+
+        # Require two-factor authentication if it has been configured or if the user is staff user.
+        if self.has_2fa_enabled(user) or user.is_staff or user.is_superuser:
+            self.stash_pending_login(request, user, **kwargs)
+            # For the social login the Require2FA middleware does not work, so check here again
+            if (
+                user.is_staff or user.is_superuser
+            ) and not self.has_2fa_enabled(user):
+                redirect_url = reverse("two-factor-setup")
```

Comment:
The `BaseRequire2FAMiddleware` middleware does not work with social login because at the time that the middleware is evaluated, the user is not logged in yet. I tried to find a fix for that in the middleware itself, but did not succeed and spent too much time trying. Checking here again is not a pretty solution, but it works.

Reply:
I'm a little bit stuck on exactly how to review this PR, honestly I don't know enough about Django AllAuth 2FA to work this out. It is not clear to me exactly what comes from the linked PR, what has been changed, and what can be removed when that PR is released. Could you maybe add some more comments to the new methods, identifying what is verbatim and what you have changed?

Reply:
I found out that my custom addition here actually breaks the login for staff users who have not set-up 2FA yet. So I removed the custom part. I now copied the linked PR as is, which means that once the linked PR has been merged, we can drop the `pre_login()`m `stash_pending_login()` and `unstash_pending_login_kwargs()`. 

The `post_login` stuff is necessary if we want to force 2FA set-up for admin users who come from the social route, because the `BaseRequire2FAMiddleware` does not appear to work with social login. I honestly think it would be fine to drop this as well since this only concerns Harm as far as I know and we can just ask him to enable 2FA, no need to force it from the Django side.

Reply:
Yes removing it sounds fine to me. Maybe on the users list page in the admin you can add the 2FA status, then site admins can enforce whatever policy they like through social controls.